drivers/storage/portworx/component/windows.go

```diff
@@ -0,0 +1,219 @@
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	WindowsComponentName     = "Windows Node"
+	WindowsDaemonSetName     = "px-csi-node-win"
+	WindowsStorageClass      = "px-csi-win"
+	WindowsDaemonSetFileName = "win.yaml"
+	kubeSystemNameSpace      = "kube-system"
+)
+
+type windows struct {
+	client        client.Client
+	isCreated     bool
+	isWindowsNode bool
+}
+
+func (w *windows) Initialize(
+	k8sClient client.Client,
+	_ version.Version,
+	scheme *runtime.Scheme,
+	recorder record.EventRecorder,
+) {
+	w.client = k8sClient
+}
+
+func (w *windows) Name() string {
+	return WindowsComponentName
+}
+
+func (w *windows) Priority() int32 {
+	return DefaultComponentPriority
+}
+
+func (w *windows) IsPausedForMigration(cluster *corev1.StorageCluster) bool {
+	return util.ComponentsPausedForMigration(cluster)
+}
+
+func (w *windows) IsEnabled(cluster *corev1.StorageCluster) bool {
+	nodeList := &v1.NodeList{}
+	err := w.client.List(context.TODO(), nodeList, &client.ListOptions{})
+	if err != nil {
+		return false
+	}
+
+	w.isWindowsNode = isWindowsNode(nodeList)
+	return w.isWindowsNode
+}
+
+func (w *windows) Reconcile(cluster *corev1.StorageCluster) error {
+	ownRef := metav1.NewControllerRef(cluster, pxutil.StorageClusterKind())
+
+	var errList []string
+
+	if err := w.createDaemonSet(WindowsDaemonSetFileName, ownRef, cluster); err != nil {
+		logrus.Errorf("error during creating %s daemonset %s ", WindowsDaemonSetName, err)
+		errList = append(errList, err.Error())
+	}
+
+	if err := w.createStorageClass(); err != nil {
+		logrus.Errorf("error during creating %s storageclass %s ", WindowsStorageClass, err)
+		errList = append(errList, err.Error())
+	}
+
+	if len(errList) > 0 {
+		return commonerrors.New(strings.Join(errList, " , "))
+	}
+	return nil
+}
+
+func (w *windows) Delete(cluster *corev1.StorageCluster) error {
+
+	ownerRef := metav1.NewControllerRef(cluster, pxutil.StorageClusterKind())
+	if err := k8s.DeleteDaemonSet(w.client, WindowsDaemonSetName, kubeSystemNameSpace, *ownerRef); err != nil {
+		return err
+	}
+	w.MarkDeleted()
+	return nil
+}
+
+func (w *windows) MarkDeleted() {
+	w.isCreated = false
+}
+
+// RegisterPortworxPluginComponent registers the PortworxPlugin component
+func RegisterWindowsComponent() {
+	Register(WindowsComponentName, &windows{})
+}
+
+func init() {
+	RegisterWindowsComponent()
+}
+
+func (w *windows) createDaemonSet(filename string, ownerRef *metav1.OwnerReference, cluster *corev1.StorageCluster) error {
+	daemonSet, err := k8s.GetDaemonSetFromFile(filename, pxutil.SpecsBaseDir())
+	if err != nil {
+		return err
+	}
+
+	daemonSet.Name = WindowsDaemonSetName
+	daemonSet.Namespace = kubeSystemNameSpace
```

Comment:
Does this need to be installed in kube-system? Can it not be installed in the cluster's namespace?

Reply:
Yes, this daemonset needs to be installed in kube-system only, it doesn't work in any other namespace, but i have changed namespace to cluster.Namespace because we have a challenge here
> Owner references work within the same namespace. So, if portworx is installed in any other namespace apart from kube-system, then this daemonset is not scheduling pods.

So this will enforce us to install portworx only in kube-system namespace while we have windows node.

Do we have any alternative solution here?

Reply:
Okay, got it.
One solution you can try is installing the DS in the kube-system namespace without the owner references. We do that for the portworx_proxy component.